### PR TITLE
#2277 - IER12 yellow fields implementation

### PR DESCRIPTION
**Pull request by dheepak-aot**
## IER12 Yellow Fields

For reference Please use the spreadsheet from General > Files >Analysis > Institution File Specs > IER12_Analysis_Output_and_Mapping_Part_1_V2.xlsx

from teams.
- [x] Updated the yellow fields in the spreadsheet with further analysis details during the implementation of the ticket. 
- [x] Created workflow calculation to get quantity of all type of dependants. (Child | Infants | Child in daycare | declared on taxes due to disability | attending post secondary)
Note: During the calculation of the age it is important to consider the following. e.g. a dependent between 12 -18 years of age is considered as child. It means dependent who is from 12 years older(right from 12th birthday) and up to  18y + 364d  old(9y-1d).

image

- [x] Update in analysis for `Spouse in school` field. As we already have a workflow field `calculatedDataPartnerStudentStudyWeeks` which corresponds to full-time post secondary study weeks of the applicant's partner during the study period of the applicant. This field (calculatedDataPartnerStudentStudyWeeks) takes input from both student application data as well as supporting user data based on the situation.

- [x] In camunda leveraged the `is defined()` function to identify the variables which are not guaranteed by the flow execution to set null value to them. Hence it is not required anymore to set null during the start of the workflow to the variables which are not guaranteed. 
Note: Null initialization was removed on those variables which were par of `configure workflow data` event. There are still variables which are initialized to null which are used in other parts of the workflow.

image

- [x] Created private method to check outstanding overaward to reduce the complexity of the method(sonar was complaining on the same)
image

- [x] Implemented the mapping for all the yellow fields
### Generated File
[IER_012_2023-10-03_08.32.57.txt](https://github.com/bcgov/SIMS/files/12795407/IER_012_2023-10-03_08.32.57.txt)

### Index in the file for yellow fields
Family dependent child quantity: 546
Family dependent daycare quantity: 549
Family dependent infant quantity: 552
Family dependent other quantity: 555
Family dependent post secondary quantity: 558
Family parent 1 flag: 567
Family parent 2 flag: 568
Parental discretionary income annual formula result: 620
Spouse in school flag: 631

### CODE REVIEW

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12-file-detail.ts [hunk 1]
```diff
@@ -81,33 +81,24 @@ export class IER12FileDetail implements IER12FileLine {
...
   dependantOtherQuantity?: number;
-  // Analysis pending for the field.
   dependantPostSecondaryQuantity?: number;
   totalDependantQuantity?: number;
   familyMembersQuantity: number;
-  // Analysis pending for the field.
-  parent1Flag?: string;
-  // Analysis pending for the field.
-  parent2Flag?: string;
+  parent1Flag?: YNFlag;
```
Comment (guru-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12-file-detail.ts:91]: 👍

File: sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn [hunk 2]
```diff
@@ -1846,9 +1831,55 @@ or (calculatedDataPartnerFedralProvincialPDReceiptCost &gt; 0)</bpmn:conditionEx
...
       </bpmn:extensionElements>
-      <bpmn:incoming>Flow_1cl7iq9</bpmn:incoming>
+      <bpmn:incoming>Flow_0i9i5pv</bpmn:incoming>
       <bpmn:outgoing>Flow_1l0g9nv</bpmn:outgoing>
     </bpmn:intermediateThrowEvent>
+    <bpmn:intermediateThrowEvent id="Event_123oime" name="calculatedDataTotalEligibleDependants">
+      <bpmn:extensionElements>
+        <zeebe:ioMapping>
+          <zeebe:output source="=if studentDataDependants = null then 0 else&#10;  count(&#10;studentDataDependants[&#10;// 0-18 years of age: eligible.&#10;date(item.dateOfBirth) &#62;= (today() - duration(&#34;P18Y&#34;))&#10;// 19-22 years of age: post-sec must be Yes or Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (today() - duration(&#34;P18Y&#34;))&#10;  and date(item.dateOfBirth) &#62;= (today() - duration(&#34;P22Y&#34;))&#10;  and (item.attendingPostSecondarySchool = &#34;yes&#34; or item.declaredOnTaxes = &#34;yes&#34;)&#10;)&#10;// &#62; 22 years: eligible if Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (today() - duration(&#34;P22Y&#34;))&#10;  and item.declaredOnTaxes = &#34;yes&#34;&#10;)&#10;]&#10;  )" target="calculatedDataTotalEligibleDependants" />
```
Comment (guru-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn:1840]: can we please change the brackets from 
image
 to 
 
image

File: sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn [hunk 3]
```diff
@@ -1846,9 +1831,55 @@ or (calculatedDataPartnerFedralProvincialPDReceiptCost &gt; 0)</bpmn:conditionEx
...
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_1q59j4l</bpmn:incoming>
+      <bpmn:outgoing>Flow_1cl7iq9</bpmn:outgoing>
+    </bpmn:intermediateThrowEvent>
+    <bpmn:intermediateThrowEvent id="Event_0s8ztj5" name="calculatedDataTotalEligibleDependants">
+      <bpmn:extensionElements>
+        <zeebe:ioMapping>
+          <zeebe:output source="=if studentDataDependants = null then 0 else&#10;  count(&#10;studentDataDependants[&#10;// 0-18 years of age: eligible.&#10;date(item.dateOfBirth) &#62;= (today() - duration(&#34;P18Y&#34;))&#10;// 19-22 years of age: post-sec must be Yes or Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (today() - duration(&#34;P18Y&#34;))&#10;  and date(item.dateOfBirth) &#62;= (today() - duration(&#34;P22Y&#34;))&#10;  and (item.attendingPostSecondarySchool = &#34;yes&#34; or item.declaredOnTaxes = &#34;yes&#34;)&#10;)&#10;// &#62; 22 years: eligible if Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (today() - duration(&#34;P22Y&#34;))&#10;  and item.declaredOnTaxes = &#34;yes&#34;&#10;)&#10;]&#10;  )" target="calculatedDataTotalEligibleDependants" />
```
Comment (guru-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn:1849]: https://github.com/bcgov/SIMS/pull/2371/files#r1344825052

File: sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn [hunk 4]
```diff
@@ -717,7 +709,7 @@
...
...   then calculatedDataTotalChildCareCost else null,&#10;    totalNonEducationalCost: if is defined(calaulatedDataTotalNonEducationalCost) &#10;    then calaulatedDataTotalNonEducationalCost else null,&#10;    dependantChildQuantity: if is defined(calculatedDataDependantChildQuantity) &#10;    then calculatedDataDependantChildQuantity else null,&#10;    dependantChildInDaycareQuantity: if is defined(calculatedDataDependantChildInDaycareQuantity)&#10;    then calculatedDataDependantChildInDaycareQuantity else null,&#10;    dependantInfantQuantity: if is defined(calculatedDataDependantInfantQuantity)&#10;    then calculatedDataDependantInfantQuantity else null,&#10;    dependantDeclaredOnTaxesQuantity: if is defined(calculatedDataDependantDeclaredOnTaxesQuantity)&#10;    then calculatedDataDependantDeclaredOnTaxesQuantity else null,&#10;    dependantPostSecondaryQuantity: if is defined(calculatedDataDependantPostSecondaryQuantity)&#10;    then calculatedDataDependantPostSecondaryQuantity else null,&#10;    partnerStudentStudyWeeks: if is defined(calculatedDataPartnerStudentStudyWeeks)&#10;    then calculatedDataPartnerStudentStudyWeeks else null&#10;  }&#10;}" target="workflowData" />
```
Comment (guru-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn:712]: Nice implementation on if is defined

File: sources/packages/backend/workflow/BPMN/camunda-8/supporting-user-information-request.bpmn [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:modeler="http://camunda.org/schema/modeler/1.0" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.4.1" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="8.1.0" camunda:diagramRelationId="422e08d6-99e6-4c1c-832e-1a661f643a3f">
```
Comment (guru-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/supporting-user-information-request.bpmn:2]: why is this file change?, just a version update?

**Review by guru-aot — COMMENTED** [review]
Nice work @dheepak-aot, just a minor comment on the old implementation, Please have a look.

File: sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn (hunk 2 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn:1840]: @guru-aot I couldn't see the problem with the brackets.

**Review by andrepestana-aot — COMMENTED** [review]
Good job. I added a comment about Guru's comment. Looks good to me.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.integration.service.ts [hunk 6]
```diff
@@ -147,8 +148,22 @@ export class IER12IntegrationService extends SFTPIntegrationBase<void> {
...
+      ierFileDetail.dependantPostSecondaryQuantity =
+        ierRecord.dependantPostSecondaryQuantity;
       ierFileDetail.totalDependantQuantity = ierRecord.totalEligibleDependents;
       ierFileDetail.familyMembersQuantity = ierRecord.familySize;
+      ierFileDetail.parent1Flag = this.convertToYNFlag(
+        ierRecord.numberOfParents > 0,
+      );
+      ierFileDetail.parent2Flag = this.convertToYNFlag(
+        ierRecord.numberOfParents === 2,
+      );
```
Comment (ann-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.integration.service.ts:166]: To refresh my mind, if there are 2 parents then, both parent 1 and 2 flags will be `Y` right?

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.processing.service.ts [hunk 7]
```diff
@@ -26,6 +26,7 @@ import { StudentAssessmentService } from "@sims/integrations/services";
 import { DisbursementOverawardService } from "@sims/services";
 import { FullTimeAwardTypes } from "@sims/integrations/models";
 import { PROVINCIAL_DEFAULT_RESTRICTION_CODE } from "@sims/services/constants";
+import { AwardOverawardBalance } from "@sims/services/disbursement-overaward/disbursement-overaward.models";
```
Comment (ann-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.processing.service.ts:29]: maybe we add it to the index file and import it from there?

File: sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn
Comment (ann-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn:1]: @dheepak-aot, I didn't get this 
image
if it is 12-18, then shouldn't it be from 12 years to 18 years - 1 day`? or is it like `12 years to 19 years - 1 day`?

File: sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn [hunk 8]
```diff
@@ -420,14 +420,6 @@
           <zeebe:output source="=null" target="studentDataTaxReturnIncome" />
           <zeebe:output source="=[]" target="parentSubprocesses" />
           <zeebe:output source="=null" target="calculatedDataParentalAssets" />
```
Comment (ann-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn:422]: shouldn't `calculatedDataParentalAssets` also be removed?

File: sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn
Comment (ann-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn:1]: Just wanted to double-check, there are more files using image
image, shouldn't we make an update or add a null check in those areas as well?

File: sources/packages/backend/workflow/BPMN/camunda-8/supporting-user-information-request.bpmn (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/supporting-user-information-request.bpmn:2]: It is updated by mistake. I will remove the file from PR.

File: sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn (hunk 4 above)
Comment (ann-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn:712]: I see, `studentDataNumberOfParents` and `calculatedDataParentalAssets` are not removed, is it intentional?

File: sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn:1840]: @guru-aot I do not want to change anything which is not part of the PR and more importantly the IDE/Editor based errors could possibly in more places.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.integration.service.ts (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.integration.service.ts:166]: @ann-aot yes as per the analysis output.

File: sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn
Comment (ann-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn:1]: looking into the logic, I think, I got what you meant in the description, its `12 years to 19 years - 1 day`

**Review by ann-aot — COMMENTED** [review]
NIce work @dheepak-aot . Some comments and questions added

File: sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn:422]: Nice catch. I see that it was set to null during initial efforts of IER 1st big chunk implementation. Will update it.

File: sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2021-2022.bpmn:1]: using the deleted - you meant `calculatedDataParentalAssets`? If so `calculatedDataParentalAssets` was working fine without null initialization before. So about that field we should be good. 

There are a bunch of fields initialized to null in the workflow which I didn't change as I don't want to add null checks to the existing code precisely outside IER scope.

File: sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/assessment-gateway.bpmn:712]: `calculatedDataParentalAssets` was missed as per one of the previous comments and `studentDataNumberOfParents` was already there initialized to null and I don't want face/validate the side effects of removing that in the story/PR. THat is why I kept it.

**Review by andrewsignori-aot — APPROVED** [review]
Nice work, looks good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @dheepak-aot

**Review by ann-aot — APPROVED** [review]
Good work @dheepak-aot 👍 Thanks for doing the changes 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks good!

**Pull request merged**